python/lsst/obs/base/cameraMapper.py

```diff
@@ -1028,36 +1029,100 @@ def _setCcdDetector(self, item, dataId, trimmed=True):
+        # result if one exists.
+        matches = set(definitions)
+        if idFilter:
+            matches.intersection_update(definitions.findAll(idFilter))
+        if filterLabel and filterLabel.hasPhysicalLabel():
+            matches.intersection_update(definitions.findAll(filterLabel.physicalLabel))
+        if filterLabel and filterLabel.hasBandLabel():
+            matches.intersection_update(definitions.findAll(filterLabel.bandLabel))
+        return matches
+
     def _setFilter(self, mapping, item, dataId):
-        """Set the filter object in an Exposure.  If the Exposure had a FILTER
-        keyword, this was already processed during load.  But if it didn't,
-        use the filter from the registry.
+        """Set the filter information in an Exposure.
+
+        The Exposure should already have had a filter loaded, but the reader
+        (in ``afw``) had to act on incomplete information. This method
+        cross-checks the filter against the data ID and the standard list
+        of filters.
 
         Parameters
         ----------
         mapping : `lsst.obs.base.Mapping`
-            Where to get the filter from.
+            Where to get the data ID filter from.
         item : `lsst.afw.image.Exposure`
             Exposure to set the filter in.
         dataId : `dict`
             Dataset identifier.
         """
-
         if not (isinstance(item, afwImage.ExposureU) or isinstance(item, afwImage.ExposureI)
                 or isinstance(item, afwImage.ExposureF) or isinstance(item, afwImage.ExposureD)):
             return
 
-        if item.getFilter().getId() != afwImage.Filter.UNKNOWN:
-            return
-
-        actualId = mapping.need(['filter'], dataId)
-        filterName = actualId['filter']
-        if self.filters is not None and filterName in self.filters:
-            filterName = self.filters[filterName]
         try:
-            item.setFilter(afwImage.Filter(filterName))
-        except pexExcept.NotFoundError:
-            self.log.warn("Filter %s not defined.  Set to UNKNOWN." % (filterName))
+            filterDefinitions = self.getGen3Instrument()().filterDefinitions
+        except NotImplementedError:
+            filterDefinitions = None
+        itemFilter = item.getFilterLabel()  # may be None
+        try:
+            idFilter = mapping.need(['filter'], dataId)['filter']
+        except dafPersist.NoResults:
+            idFilter = None
+
+        if filterDefinitions:
+            definitions = self._resolveFilters(filterDefinitions, idFilter, itemFilter)
+            self.log.debug("Matching filters for id=%r and label=%r are %s.",
+                           idFilter, itemFilter, definitions)
+            if len(definitions) == 1:
+                newLabel = list(definitions)[0].makeFilterLabel()
+                if newLabel != itemFilter:
+                    item.setFilterLabel(newLabel)
+            elif definitions:
```

Comment:
Should this be `len(definitions) > 1` to be explicit?

Reply:
I was told that's Unpythonic, and that `if definitions` is the preferred style?